### ISSUE TITLE
not increment pInfo.Attempts when the scheduling fails with Error status

### ISSUE THREAD

**Issue by sanposhiho** [kubernetes/kubernetes#128744]
The current `pInfo.Attempts` is simple; when the scheduler `Pop()` the pod, it increments `pInfo.Attempts`.
And, also this `pInfo.Attempts` is used to calculate the backoff time.

There are basically two failures that the scheduling cycle could make; Unschedulable or Error.
- Unschedulable: the Pod is unschedulable; e.g., no node has enough resource, nodeaffinity, etc.
- Error: the Pod cannot be scheduled because of unexpected errors;  e.g., kube-apiserver is unstable, network issue, etc.

Given [the backoff is a penalty of wasting the scheduling cycle](https://github.com/kubernetes/kubernetes/blob/8fe10dc378b7cc3b077b83aef86622e1019302d5/pkg/scheduler/framework/interface.go#L205-L207), we shouldn't increment `Attempts` when the Pod comes back to queue with Error status
so that the backoff time doesn't get longer by those errors.

As the actual example, what could happen now is:
1. The scheduler fails to schedule all the pods because the cluster is under the networking issue.
2. `pInfo.Attempts` in all the Pods gets bigger and bigger meanwhile. 
3. The network issue is resolved and some Pods go scheduled.
4. But, then if some Pods fail, they have a large `pInfo.Attempts`, and hence gets super long backoff time, which is unfair because the most of their `Attempts` are failed not by the scheduling reason, but by the network issue.

/kind feature
/assign
/sig scheduling

**Comment by haosdent** [comment]
> Error: the Pod cannot be scheduled because of unexpected errors; e.g., kube-apiserver is unstable, network issue, etc.

Actually I feel we may need to back off when the kube-apiserver is unstable as well, to avoid overloading kube-apiserver

**Comment by sanposhiho** [comment]
This change doesn't change how much kube-scheduler putting load to kube-apiserver because when the scheduling cycle fails with Error, we skip backoff already.
So, the problem here is **after** kube-apiserver is recovered; like the above my example scenario said, once it's recovered, kube-scheduler starts to schedule Pods as usual, but if some unfortunate Pods get unschedulable, they get toooo long backoff time because of kube-apiserver downtime

**Comment by dom4ha** [comment]
> This change doesn't change how much kube-scheduler putting load to kube-apiserver because when the scheduling cycle fails with Error, we skip backoff already. 

Can you point to a specific place where it happens?

**Comment by sanposhiho** [comment]
> I agree that we can skip the backoff queue, but I'd suggest to have an exponential backoff on api calls first.

Yes, we have to do both ideas on my comment. Especially cannot skip the first one if we do the second one.

> It's especially important in case of our asynchronous scenarios, where the exponential backoff would not block the scheduling itself, but potentially a set of concurrent requests.

It reminds me of a thing, which I discussed with someone somewhere..
So, currently, we have `Parallelism` parameter in the scheduler config, but we don't respect it in some parts like the binding cycle goroutines. Maybe we should, and then starting a new binding cycle would be blocked when many binding cycles are retrying their API calls. But, obviously, we have to pay attention to degradation if we do this.

**Comment by haosdent** [comment]
> I agree that we can skip the backoff queue, but I'd suggest to have an exponential backoff on api calls first.

Agree with this as well. Thanks @sanposhiho 's clarification

**Comment by dom4ha** [comment]
It seems to me that adding exponential backoff to the api calls would be similar (in principle) to the backoff queue we are trying to eliminate. In both cases Pod would have to pay a price of the errors they encounter.

So, since we already have a backoff queue, maybe we just could apply different values to the error cases compared to the unschedulable cases?

**Comment by macsko** [comment]
I think, using backoff on API calls can be even worse for apiserver than not doing it. Suppose you get failed API calls in asynchronous parts, such as binding. Won't those goroutines stack up, and won't the scheduler add new pods to that stack? It would result in even more API calls (increase the load) and more goroutines than before. 

What if we try another solution, such as we could have two counters for the pods in the queue? One for error and one for unscheduled attempts. If a pod encounters errors during scheduling/binding, it will get a backoff based on the error counter, but if the errors stop occurring, the pod will only be penalized by the unscheduled counter. We could even try to adjust the backoff policy for the error counter, if necessary, in the future.

**Comment by sanposhiho** [comment]
I like @macsko 's idea. Will update #128748 (or maybe create a separate PR, depending on the diff size vs how much each might conflict)
